airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination/buffered_stream_consumer/BufferedStreamConsumer.java

```diff
@@ -192,12 +203,14 @@ protected void close(boolean hasFailed) throws Exception {
     }
 
     try {
-      onClose.accept(hasFailed);
-      // todo (cgardens) - For now we are using this conditional to maintain existing behavior. When we
-      // enable checkpointing, we will need to get feedback from onClose on whether any data was persisted
-      // or not. If it was then, the state message will be emitted.
-      if (!hasFailed && lastCommittedState != null) {
-        outputRecordCollector.accept(lastCommittedState);
+      // if any state message flushed that means we can still go for at least a partial success. if none
+      // was emitted, if there were still no failures, then we can still succeed. the latter case is full
+      // refresh.
+      onClose.accept(lastFlushedState == null && hasFailed);
```

Comment:
I've re-read this line like 5 times to map out the boolean table in my head. Do you mind unfurling this e.g: 

```
if (lastFlushedState == null) { 
  onClose.accept(hasFailed)
} else { 
  outputRecordCollector.accept(lastFlushedState)
  onClose.accept( false ) 
} 
```

Reply:
hahaha. no i don't mind at all. it took me a long time to write this expression (which should have been a sign that it was too hard to think about)

Reply:
```
if (lastFlushedState == null) { 
  onClose.accept(hasFailed)
} else { 
  
  onClose.accept( false ) 
} 

if (lastFlushedState != null) {
   outputRecordCollector.accept(lastFlushedState);
}

```

Reply:
why separate them?

Reply:
Notes from talking with davin: 
* hasFailed
* clarify the lastFlushedState concept (especially clarifying in the context of the copy strategy)

Reply:
also need to add re throwing the exception on close?

Reply:
Spoke offline: went through code and confirmed this works for Copy.

One note here to rename `hasFailed` to something closer to `commit` or `discard` since this no longer represents failure.

Reply:
@sherifnada they need to be separate because the condidiontal is not the same. one is `lastFlushedState = null` and one is `lastFlushedState != null`

Reply:
couldn't it go in the `else` clause?